Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,409 @@
+
+### Test coverage
+
+As intrinsic support is added test coverage must be extended to provide basic testing
+
+## Half precision floating point
+
+This document will refer to half precision floating point as `Half`.
+
++ Machine learning and Artificial intelligence often use `Half` type to simplify storage and improve processing time.
++ CoreCLR and `CIL` in general do not have general support for a `Half` type
++ There is an open request to expose `Half` intrinsics
++ There is an outstanding proposal to add `System.Half` to support this request https://github.com/dotnet/corefx/issues/25702
++ Implementation of `Half` features will be adjusted based on
+  + Implementation of the `System.Half` proposal
+  + Availability of supporting hardware (extensions)
+  + General language extensions supporting `Half`
+
+### ARMv8 Half precision support
+
+ARMv8 baseline support for `Half` is limited.  The following operations are supported
+
++ Loads and Stores
++ Conversion to/from `Float`
++ Widening from `Vector128<Half>` to two `Vector128<Float>`
++ Narrowing from two `Vector128<Float>` to `Vector128<Half>`
+
+Recent extensions add support for
+
++ General operations on `Half` types
++ Vector operations on `Half` types
+
+### `Half` and ARM64 ABI
+
+Any complete `Half` implementation must conform to the `ARM64 ABI`.
+
+The proposed `System.Half` type must be treated as a floating point type for purposes of the ARM64 ABI
+
+As an argument it must be passed in a floating point register.
+
+As a structure member, it must be treated as a floating point type and enter into the HFA determination logic.
+
+Test cases must be written and conformance must be demonstrated.
+
+## Scalable Vector Extension Support
+
+`SVE`, the Scalable Vector Extension introduces its own complexity.
+
+The extension
+
++ Creates a set of `Z0-Z31` scalable vector registers.  These overlay existing vector registers.  Each scalar vector register has a platform specific length
+  + Any multiple of 128 bits up to 2048 bits
++ Creates a new set of `P0-P15` predicate registers.  Each predicate register has a platform specific length which is 1/8th of the scalar vector length.
++ Add an extensive set of instructions including complex load and store operations.
++ Modifies the ARM64 ABI.
+
+Therefore implementation will not be trivial.
+
++ Register allocator will need changes to support predicate allocation
++ SIMD support will face similar issues
++ Open issue: Should we use `Vector<T>`, `Vector128<t>, Vector256<t>, ... Vector2048<T>`, `SVE<T>` ... in user interface design?
+  + Use of `Vector128<t>, Vector256<t>, ... Vector2048<T>` is current default proposal.
+Having 16 forms of every API may create issues for framework and client developers.
+However generics may provide some/sufficient relief to make this acceptable.
+  + Use of `Vector<T>` may be preferred if SVE will not be used for `FEATURE_SIMD`
+  + Use of `SVE<T>` may be preferred if SVE will not be used for `FEATURE_SIMD`
```

Comment:
I'm not sure I understand this - both "may be preferred if SVE will not be used for `FEATURE_SIMD`". Should the second be "will be used"?

Reply:
The first one is wrong.  Late edit introduced a typo.

If FEATURE_SIMD is updated to be implemented using SVE ISA, then `Vector<T>` is likely preferred.  Otherwise, a new variable length vector type is preferred (i.e. `SVE<T>`).

I will fix as well as incorporate comments from the other review.

Reply:
I think this warrants further discussion when the decision to begin implementing SVE starts happening. There are several issues with variable length registers (as was encountered with `System.Numerics.Vector<T>`) and there may be a different/better solution available (such as explicitly sized registers each with their own `IsSupported` flag).

My understanding of SVE, from reading the supplement (DDI 0584) is that the underlying register size is variable from 128-bits up to 2048-bits, in 128-bit increments. However, a machine which supports 2048-bits will also support accessing and consuming the registers as if they were 1024, 512, 384-bits, etc.

Additionally, the registers are shared with the SIMD/FP registers, such that the SIMD/FP registers are just the first 128-bits of the SVE registers.